lib/pure/collections/sequtils.nim

```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
-  type OutType = typeof(items(s))
-  when compiles(s.len):
-    block:
-      evalOnceAs(s2, s, compiles((let _ = s)))
-      var i = 0
-      var result = newSeq[OutType](s2.len)
-      for it in s2:
-        result[i] = it
-        i += 1
-      result
-  else:
-    var result: seq[OutType] = @[]
-    for it in s:
-      result.add(it)
-    result
+template toSeqBuiltin[IX, T](arg: array[IX, T]): seq[T] = @arg
 
-template toSeq2(iter: iterator): untyped =
-  # overload for iterator
-  evalOnceAs(iter2, iter(), false)
-  when compiles(iter2.len):
-    var i = 0
-    var result = newSeq[typeof(iter2)](iter2.len)
-    for x in iter2:
-      result[i] = x
-      inc i
-    result
+template toSeqBuiltin[T](arg: openArray[T]): seq[T] = @arg
+
+template toSeqBuiltin(arg: string): seq[char] = @arg
+
+template toSeqBuiltin[T](arg: seq[T]): seq[T] = arg
+
+proc toSeqBuiltin[T](arg: set[T]): seq[T] {.inline.} =
+  for it in arg:
+    result.add it
+
+proc toSeqBuiltin(a: cstring): seq[char] {.inline.} =
+  let len = a.len
+  result.setLen len
+  var i = 0
+  while i < len:
+    result[i] = a[i]
+    inc i
+
+proc toSeqBuiltin[E: enum](t: typedesc[E]): seq[E] =
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
+  let tmpSym = genSym(nskVar)
+  result = nnkStmtListExpr.newTree
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
+  result.add quote do:
+    for it in `arg`:
+      `tmpSym`.add it
+  result.add tmpSym
+
+template toSeqBuiltin[T: not proc](arg: T): untyped =
+  # fallback for ``IntSet`` and ``HashSet``. Explicit overload would
+  # be better, but this would require an additional import here.
+  var tmp: seq[typeof(items(arg), typeOfIter)]
+  for it in arg:
+    tmp.add it
+  tmp
+
+macro toSeqImpl(arg: typed): untyped =
+  let iteratorCall = arg[1]
+  let iteratorSym = iteratorCall[0]
+  if iteratorCall.len == 2 and iteratorSym.eqIdent("items") and iteratorSym.owner.owner.eqIdent("stdlib"):
+    let iteratorArg = iteratorCall[1]
+    result = newCall(bindSym"toSeqBuiltin", iteratorArg)
   else:
-    type OutType = typeof(iter2())
-    var result: seq[OutType] = @[]
-    when compiles(iter2()):
-      evalOnceAs(iter4, iter, false)
-      let iter3 = iter4()
-      for x in iter3():
-        result.add(x)
-    else:
-      for x in iter2():
-        result.add(x)
-    result
+    var typ = iteratorCall.getTypeInst()
+    # `lent` is broken feature. and this special handling to get rid of lent is why it is broken.
+    if typ.kind == nnkBracketExpr and typ.len == 2 and typ[0].eqIdent("lent"):
+      # TODO this is not guaranteed to be the acutal system `lent` symbol.
+      typ = typ[1]
```

Comment:
You can use this to get the return type of any iterable expression:

```nim
template iterRetTyp(u: untyped): untyped =
  proc inferer(): auto {.genSym.} =
    for it in u:
      return it

  typeof(inferer())

iterator test[T](): T =
  yield default(T)

echo iterRetTyp(test[int]())
echo iterRetTyp(['a', 'b', 'c'].items)
```

With this you basically hand the job to the compiler, so it's always correct.

Reply:
I think you can use `typeKind` to identify the type of `lent` and/or `var` to skip as well.

Reply:
I've written extensive macros. I've written Nim to glsl compilers like in OpenGL sandbox. I've never ever used `typeKind` for anything, and would like to keep it that way.

Reply:
You are going into `typed` land and the best way to be correct is to ask the compiler. It's one of the parts in typed macro to be solid-ish and trustable.

Reply:
It is not about typed vs untyped. It is about the fact that `typeKind` feels 100% redundant to `getTypeInst.kind`. It is not obvious why a separate proc with its own enum type is necessary here, when all of this information is already present in the ast.

Maybe you are right in terms of getting rid of the `lent` flag (And I will change my PR accordingly). But that doesn't change my opinion on the redundancy of this feature. After all this information is contained in the type ast, therefore it could be eleminated.

Reply:
I just tested `typeKind`, it returns `ntyUnused2`. In other words, no `typeKind` can't be used here.

Reply:
@alaviss regarding `iterRetTyp` I've tested it. and it works. But I'd rather not use it here. The major reason for it is, I am in macro land, I should be able to just ask the compiler "gimme the return type of this iterator", I should not need to rely of crutches like you suggested to get my type. And I do it, I ask the compiler to get the type, but the compiler also puts in garbage into my answer and that is a problem that needs to be fixed. In order for problem to be fixed, they need to be visible. I just can't fix it within this PR.

Reply:
Can you check if comparing directly with `bindSym"lent"` works then? That'd be a lot more robust.